### ISSUE TITLE
Optionally describe a channel as projection over a subset of messages from a message source

### ISSUE THREAD

**Issue by sammy-hughes** [asyncapi/spec#923]
The async-api spec provides facility to describe consumer attention in the abstract, but a specific application as consumer might attend to a subset of events on a specific channel. There may even be distinct behaviors for overlapping subsets of messages on a given queue or event-source.

Even in the 3.0.0 release, describing such same-source-different-behavior workflows in async-api blurs the line between specifying filtering conditions and validation concerns. It is proposed that some means of describing a channel as a filtered subset of the actual source would help keep filtering and validation concerns separate.

Use the [examples/application-headers.yaml](https://github.com/asyncapi/spec/blob/master/examples/application-headers.yml) specification for this problem. Suppose I want to add a service which identifies failing lights and publishes an event to a different channel, to cause a technician to be dispatched to repair lights which measure under an arbitrary threshold.

A possible naive solution to that problem is to describe a message which is like lightMeasured, but which would exclude any value for `lumens` not between 0 and 39. Any hypothetical measurement under 40 lumens is interpreted as an equipment failure.
```YAML
channels:
  smartylighting/streetlights/1/0/event/{streetlightId}/lighting/measured:
    ...
    publish:
      ...
      message:
        $ref: '#/components/messages/lightMeasured'
  smartylighting/streetlights/1/0/event/{streetlightId}/lighting/failed:
    ...
    publish:
      ...
      message:
        $ref: '#/components/messages/lightFailed'

components:
  messages:
    lightMeasured:
      ...
    lightFailed:
      allOf:
      - $ref: "#/components/messages/lightMeasured"
      - name: streetlightFailed
        title: Streetlight Failed
        summary: Streetlight is not currently illuminating above low-light conditions
        payload:
          $ref: "#/components/schemas/lightBelowAmbientPayload"
    
  schemas:
    lightMeasuredPayload:
      ...
    lightBelowAmbientPayload:
      type: object
      properties:
        lumens:
          type: integer
          maximum: 39
          description: light measurement does not exceed ambient lighting
    ...
```
This version of the application-headers example can now describe a single event-stream as two separate channels, with slightly different inclusion criteria. The intended interpretation is that the second channel is a projection on the first, having only messages that describe broken equipment.

Unfortunately, this version is ambiguous as to whether an implementation should report an error for any message on `smartylighting/streetlights/1/0/event/{streetlightId}/lighting/failed` which fails validation for the given schema. Such ambiguity in a specification document is self-defeating, so a better mechanism is needed.

It is proposed that the async-api specification be modified to provide some facility to describe messages which are still valid, but which are not relevant to an application. The implementation behavior that this describes should expect to regard such irrelevant messages as valid, and should not produce an error upon encountering such messages. This proposal does not stipulate any further behavioral prescription. 

In contrast to the naive approach previously described, this separates any message filtering from the message validation.

If this proposal is found to be agreeable, it could be added to the async-api specification as a property for a channel object as shown below. The choice of property name for this facility is acknowledged to be quite important, but this proposal is not opinionated as to that name; the choice of "suppress" is arbitrary.
```YAML
"http://asyncapi.com/definitions/3.0.0/channel.json":
  "$id": "http://asyncapi.com/definitions/3.0.0/channel.json"
  type: object
  additionalProperties: false
  patternProperties:
    "^x-[\\w\\d\\.\\x2d_]+$":
    "$ref": "http://asyncapi.com/definitions/3.0.0/specificationExtension.json"
  properties:
    address:
      type:
        oneOf:
        - string
        - null
      description: |
        An optional string representation of this channel's address. The address is typically the "topic name", "routing key", 
        "event type", or "path". When `null` or absent, it MUST be interpreted as unknown. This is useful when the address is 
        generated dynamically at runtime or can't be known upfront. It MAY contain Channel Address Expressions.
    messages:
      "$ref": "http://asyncapi.com/definitions/3.0.0/channelMessages.json"
    suppress:
      "$ref": "http://asyncapi.com/definitions/3.0.0/message.json"
    parameters:
      "$ref": "http://asyncapi.com/definitions/3.0.0/parameters.json"
    title:
      type: string
      description: A human-friendly title for the channel.
    summary:
      type: string
      description: A brief summary of the channel.
    description:
      type: string
      description: A longer description of the channel. CommonMark is allowed.
    servers: 
      type: array
      description: The references of the servers on which this channel is available. If absent or empty then this channel must be available on all servers.
      items: 
        "$ref": "http://asyncapi.com/definitions/3.0.0/Reference.json"
      uniqueItems: true
    tags:
      type: array,
      items:
        oneOf:
        - "$ref": "http://asyncapi.com/definitions/3.0.0/Reference.json"
        - "$ref": "http://asyncapi.com/definitions/3.0.0/tag.json"
      uniqueItems: true
    externalDocs:
      oneOf:
      - "$ref": "http://asyncapi.com/definitions/3.0.0/Reference.json"
      - "$ref": "http://asyncapi.com/definitions/3.0.0/externalDocs.json"
    bindings:
      "$ref": "http://asyncapi.com/definitions/3.0.0/channelBindingsObject.json"
```

The proposed behavior would also make sense as a property of some/all protocol channel bindings, instead of being a direct property of a channel. Async-api 3.0.0 adds significantly to the bindings schemas, and that could mean a high number of individual locations in the specification would be changed, if that version of the proposal is found to be agreeable.

**Comment by sammy-hughes** [comment]
In retrospect, this seems like one of a handful of shortcomings for async-api, around things like projections and aggregations. There isn't really language as part of async-api to describe relationships to an event-series which describe less-than-exactly-and-every handling.

**Comment by dalelane** [comment]
I've been playing with describing stream processing using AsyncAPI - something that I think is similar to what you're describing.

The way I've been approaching it is to describe the applications - the stream processors - rather than the topics.

For example, imagine I have a topic called INPUT, and a stream processor filtering these to a new topic called OUTPUT.

I could have an AsyncAPI doc that describes the application that is producing events to the INPUT topic. And I could have an AsyncAPI doc describing the stream processor that is consuming from INPUT and producing to OUTPUT. 

Taken as a whole, the collection of AsyncAPI docs describe my overall event driven solution architecture - and could be used to infer the upstream provenance of events on the OUTPUT topic.

I think that's what you're looking to achieve, am I right?

**Comment by sammy-hughes** [comment]
I get that I can specify INPUT and OUTPUT streams, but I can't describe the logic whereby INPUT elements are transformed to be elements of OUTPUT, or what my acceptance criteria might be for INPUT elements to pass to OUTPUT.

This proposal is after language to describe splitting, filtering, and other routing layers in terms of async-api notation. I could describe such layers according to name-based convention or I could extend the async-api spec, but neither would be validated by the existing async-api toolchain. Without reimplementing significant parts of the toolchain, I'm not able to write a document which has the level of detail that I believe necessary for my intended use.

The proposal I made was for a step in the direction of this level of specificity.

This really only matters if intending to procedurally generate robust services, nuanced services. Some amount of boilerplate is a reasonable code-generation target now, but I can't think of any service I've worked on that would be fine as-is, when generated from the existing async-api specification terms.

I recognize that most will not be intending to generate a service from the specification, rather simply describing an existing or proposed service, but those use-cases are also served by something like what I proposed. Transformation and filtering logic is generally what takes the bulk of service planning. So despite my lack-of-capability statement applying to a niche case, I believe such language would be of broader interest beyond this niche.
